### ISSUE TITLE
kubectl set image fails in minikube when image is present in its docker (hello minikube tutorial)

### ISSUE THREAD

**Issue by grossws** [kubernetes/minikube#1721]
BUG REPORT, please:

**Minikube version** (use `minikube version`):
`minikube version: v0.20.0`

**Environment**:
- **OS**: Arch Linux
- **VM Driver**: kvm
- **ISO version**: v0.20.0
- **kubectl version**: 
```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2+43a9be4", GitCommit:"43a9be4", GitTreeState:"clean", BuildDate:"2017-05-16T16:36:05Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.4", GitCommit:"d6f433224538d4f9ca2f7ae19b252e6fcb66a3ae", GitTreeState:"dirty", BuildDate:"2017-06-22T04:31:09Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

**What happened**:
When going through [hello-minikube](https://kubernetes.io/docs/tutorials/stateless-application/hello-minikube/) tutorial step with `set image` fails:
```bash
-> % kubectl set image deployment/hello-node hello-node=hello-node:v2
error: unable to resolve image "hello-node:v2" for container "hello-node": unable to resolve  "hello-node:v2"
```

**What you expected to happen**:
Image in deployment should be updated:
```bash
[root@kube-test mk]# kubectl set image deployment/hello-node hello-node=hello-node:v2
deployment "hello-node" image updated
```

**How to reproduce it** (as minimally and precisely as possible):
Follow hello-minikube tutorial:
- `minikube start`
- `kubectl config use-context minikube`
- get [`server.js`](https://raw.githubusercontent.com/kubernetes/kubernetes.github.io/master/docs/tutorials/stateless-application/server.js) and [`Dockerfile`](https://raw.githubusercontent.com/kubernetes/kubernetes.github.io/master/docs/tutorials/stateless-application/Dockerfile)
- `eval $(minikube docker-env)`
- `docker build -t hello-node:v1 .`
- `kubectl run hello-node --image=hello-node:v1 --port=8080`
- update `server.js` (like editing message in `response.end(...)`), rebuild image with `docker build -t hello-node:v2 .`
- try to update image with `kubectl set image deployment/hello-node hello-node=hello-node:v2`

**Anything else do we need to know**:
- it works fine with `kubeadm`-bootstrapped environment on CentOS 7 (with `calico` network) and same commands as for `minikube`;
- image `hello-node:v2` is present in `minikube` VM;
- deployment cat be updated with new image via `kubectl edit deployment/hello-node` or via dashboard's yaml editor;
- `imagePullPolicy` == `IfNotPresent` for this deployment:
```yaml
#-> % kubectl get deployment/hello-node -o yaml  
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
  creationTimestamp: 2017-07-19T18:19:28Z
  generation: 1
  labels:
    run: hello-node
  name: hello-node
  namespace: default
  resourceVersion: "105016"
  selfLink: /apis/extensions/v1beta1/namespaces/default/deployments/hello-node
  uid: cd87afbc-6cae-11e7-b1c5-5254001164ac
spec:
  replicas: 1
  selector:
    matchLabels:
      run: hello-node
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        run: hello-node
    spec:
      containers:
      - image: hello-node:v1
        imagePullPolicy: IfNotPresent
        name: hello-node
        ports:
        - containerPort: 8080
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: 2017-07-19T18:19:28Z
    lastUpdateTime: 2017-07-19T18:19:28Z
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  observedGeneration: 1
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
```

Originally posted to kubernetes/kubernetes.github.io#4416

**Comment by r2d4** [comment]
I wasn't able to reproduce with minikube v0.20.0 or at HEAD and a 1.6.0-kubectl.

The only thing I can think of is that the docker daemon is not actually pointing to minikube.  Can you ensure that `docker images` returns 

```
$ docker images
REPOSITORY                                             TAG                 IMAGE ID            CREATED              SIZE
hello-node                                             v2                  1b6651c9f72d        About a minute ago   655.4 MB
hello-node                                             v1                  31a2041fc5bb        2 minutes ago        655.4 MB
gcr.io/google_containers/kubernetes-dashboard-amd64    v1.6.1              71dfe833ce74        9 weeks ago          134.4 MB
gcr.io/google_containers/k8s-dns-sidecar-amd64         1.14.2              7c4034e4ffa4        10 weeks ago         44.5 MB
gcr.io/google_containers/k8s-dns-kube-dns-amd64        1.14.2              ca8759c215c9        10 weeks ago         52.36 MB
gcr.io/google_containers/k8s-dns-dnsmasq-nanny-amd64   1.14.2              e5c335701995        10 weeks ago         44.84 MB
gcr.io/google-containers/kube-addon-manager            v6.4-beta.1         85809f318123        4 months ago         127.2 MB
node                                                   6.9.2               faaadb4aaf9b        7 months ago         655.4 MB
gcr.io/google_containers/pause-amd64                   3.0                 99e59f495ffa        14 months ago        746.9 kB
``` 
from the same shell you're running kubectl?  `$ eval $(minikube docker-env)` will only affect the shell its run in.

**Comment by r2d4** [comment]
You might also further debug with 
`kubectl set image --v 10 deployment/hello-node hello-node=hello-node:v2`
or 
`minikube logs` for debugging the server.

**Comment by grossws** [comment]
> from the same shell you're running kubectl? $ eval $(minikube docker-env) will only affect the shell its run in.

Of course, `eval` was called on same `zsh` shell as `docker build`. Just to be sure:
```bash
-> % minikube ssh docker images | column -t
REPOSITORY                                            TAG          IMAGE         ID  CREATED  SIZE
hello-node                                            v2           f145275015ba  26  hours    ago   660.3  MB
hello-node                                            v1           3745b7a1125f  27  hours    ago   660.3  MB
<none>                                                <none>       b89e1427f5be  27  hours    ago   660.3  MB
node                                                  7            b4513a945e14  8   days     ago   660.3  MB
gcr.io/google_containers/kubernetes-dashboard-amd64   v1.6.1       71dfe833ce74  9   weeks    ago   134.4  MB
gcr.io/google_containers/k8s-dns-sidecar-amd64        1.14.2       7c4034e4ffa4  10  weeks    ago   44.5   MB
gcr.io/google_containers/k8s-dns-kube-dns-amd64       1.14.2       ca8759c215c9  10  weeks    ago   52.36  MB
gcr.io/google_containers/k8s-dns-dnsmasq-nanny-amd64  1.14.2       e5c335701995  10  weeks    ago   44.84  MB
gcr.io/google-containers/kube-addon-manager           v6.4-beta.1  85809f318123  4   months   ago   127.2  MB
gcr.io/google_containers/pause-amd64                  3.0          99e59f495ffa  14  months   ago   746.9  kB
```
`docker images` in same console returns same result.

```bash
-> % kubectl set image --v 10 deployment/hello-node hello-node=hello-node:v2
I0721 00:29:47.408872   30854 loader.go:354] Config loaded from file /home/gross/.kube/config
I0721 00:29:47.409853   30854 cached_discovery.go:112] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/servergroups.json
I0721 00:29:47.410083   30854 cached_discovery.go:112] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/servergroups.json
I0721 00:29:47.410207   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/authentication.k8s.io/v1/serverresources.json
I0721 00:29:47.410283   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/authentication.k8s.io/v1beta1/serverresources.json
I0721 00:29:47.410321   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/authorization.k8s.io/v1/serverresources.json
I0721 00:29:47.410372   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/authorization.k8s.io/v1beta1/serverresources.json
I0721 00:29:47.410419   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/autoscaling/v1/serverresources.json
I0721 00:29:47.410468   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/autoscaling/v2alpha1/serverresources.json
I0721 00:29:47.410520   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/batch/v1/serverresources.json
I0721 00:29:47.410544   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/batch/v2alpha1/serverresources.json
I0721 00:29:47.410570   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/certificates.k8s.io/v1beta1/serverresources.json
I0721 00:29:47.410614   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/extensions/v1beta1/serverresources.json
I0721 00:29:47.410665   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/policy/v1beta1/serverresources.json
I0721 00:29:47.410692   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/rbac.authorization.k8s.io/v1beta1/serverresources.json
I0721 00:29:47.410715   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/rbac.authorization.k8s.io/v1alpha1/serverresources.json
I0721 00:29:47.410736   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/settings.k8s.io/v1alpha1/serverresources.json
I0721 00:29:47.410757   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/storage.k8s.io/v1beta1/serverresources.json
I0721 00:29:47.410778   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/storage.k8s.io/v1/serverresources.json
I0721 00:29:47.410805   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/apps/v1beta1/serverresources.json
I0721 00:29:47.410879   30854 cached_discovery.go:70] returning cached discovery info from /home/gross/.kube/192.168.42.17_8443/v1/serverresources.json
I0721 00:29:47.411763   30854 round_trippers.go:299] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: kubectl/v1.5.2+43a9be4 (linux/amd64) kubernetes/43a9be4" https://192.168.42.17:8443/apis/extensions/v1beta1/namespaces/default/deployments/hello-node
I0721 00:29:47.433250   30854 round_trippers.go:318] GET https://192.168.42.17:8443/apis/extensions/v1beta1/namespaces/default/deployments/hello-node 200 OK in 21 milliseconds
I0721 00:29:47.433272   30854 round_trippers.go:324] Response Headers:
I0721 00:29:47.433277   30854 round_trippers.go:327]     Content-Type: application/json
I0721 00:29:47.433281   30854 round_trippers.go:327]     Content-Length: 1366
I0721 00:29:47.433285   30854 round_trippers.go:327]     Date: Thu, 20 Jul 2017 21:29:47 GMT
I0721 00:29:47.433315   30854 request.go:905] Response Body: {"kind":"Deployment","apiVersion":"extensions/v1beta1","metadata":{"name":"hello-node","namespace":"default","selfLink":"/apis/extensions/v1beta1/namespaces/default/deployments/hello-node","uid":"cd87afbc-6cae-11e7-b1c5-5254001164ac","resourceVersion":"138349","generation":5,"creationTimestamp":"2017-07-19T18:19:28Z","labels":{"run":"hello-node"},"annotations":{"deployment.kubernetes.io/revision":"3"}},"spec":{"replicas":1,"selector":{"matchLabels":{"run":"hello-node"}},"template":{"metadata":{"creationTimestamp":null,"labels":{"run":"hello-node"}},"spec":{"containers":[{"name":"hello-node","image":"hello-node:v1","ports":[{"containerPort":8080,"protocol":"TCP"}],"resources":{},"terminationMessagePath":"/dev/termination-log","terminationMessagePolicy":"File","imagePullPolicy":"IfNotPresent"}],"restartPolicy":"Always","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFirst","securityContext":{},"schedulerName":"default-scheduler"}},"strategy":{"type":"RollingUpdate","rollingUpdate":{"maxUnavailable":1,"maxSurge":1}}},"status":{"observedGeneration":5,"replicas":1,"updatedReplicas":1,"readyReplicas":1,"availableReplicas":1,"conditions":[{"type":"Available","status":"True","lastUpdateTime":"2017-07-19T18:19:28Z","lastTransitionTime":"2017-07-19T18:19:28Z","reason":"MinimumReplicasAvailable","message":"Deployment has minimum availability."}]}}
F0721 00:29:47.435416   30854 helpers.go:116] error: unable to resolve image "hello-node:v2" for container "hello-node": unable to resolve  "hello-node:v2"
```

```bash
-> % minikube logs
Jul 20 21:30:07 minikube localkube[2941]: I0720 21:30:07.918960    2941 operation_generator.go:597] MountVolume.SetUp succeeded for volume "kubernetes.io/secret/3ef64b0e-6cae-11e7-b1c5-5254001164ac-default-token-w4bwz" (spec.Name: "default-token-w4bwz") pod "3ef64b0e-6cae-11e7-b1c5-5254001164ac" (UID: "3ef64b0e-6cae-11e7-b1c5-5254001164ac").
Jul 20 21:30:07 minikube localkube[2941]: I0720 21:30:07.919540    2941 operation_generator.go:597] MountVolume.SetUp succeeded for volume "kubernetes.io/configmap/3ef64b0e-6cae-11e7-b1c5-5254001164ac-kube-dns-config" (spec.Name: "kube-dns-config") pod "3ef64b0e-6cae-11e7-b1c5-5254001164ac" (UID: "3ef64b0e-6cae-11e7-b1c5-5254001164ac").
Jul 20 21:30:19 minikube localkube[2941]: I0720 21:30:19.882816    2941 operation_generator.go:597] MountVolume.SetUp succeeded for volume "kubernetes.io/secret/3edb91c8-6cae-11e7-b1c5-5254001164ac-default-token-w4bwz" (spec.Name: "default-token-w4bwz") pod "3edb91c8-6cae-11e7-b1c5-5254001164ac" (UID: "3edb91c8-6cae-11e7-b1c5-5254001164ac").
Jul 20 21:30:35 minikube localkube[2941]: I0720 21:30:35.959924    2941 operation_generator.go:597] MountVolume.SetUp succeeded for volume "kubernetes.io/secret/8fd018a0-6d52-11e7-b1c5-5254001164ac-default-token-v2mq2" (spec.Name: "default-token-v2mq2") pod "8fd018a0-6d52-11e7-b1c5-5254001164ac" (UID: "8fd018a0-6d52-11e7-b1c5-5254001164ac").
Jul 20 21:31:19 minikube localkube[2941]: I0720 21:31:19.885736    2941 operation_generator.go:597] MountVolume.SetUp succeeded for volume "kubernetes.io/secret/3ef64b0e-6cae-11e7-b1c5-5254001164ac-default-token-w4bwz" (spec.Name: "default-token-w4bwz") pod "3ef64b0e-6cae-11e7-b1c5-5254001164ac" (UID: "3ef64b0e-6cae-11e7-b1c5-5254001164ac").
Jul 20 21:31:19 minikube localkube[2941]: I0720 21:31:19.902586    2941 operation_generator.go:597] MountVolume.SetUp succeeded for volume "kubernetes.io/configmap/3ef64b0e-6cae-11e7-b1c5-5254001164ac-kube-dns-config" (spec.Name: "kube-dns-config") pod "3ef64b0e-6cae-11e7-b1c5-5254001164ac" (UID: "3ef64b0e-6cae-11e7-b1c5-5254001164ac").
```

The only severe difference is that I have `kubectl` 1.5.2 and it seems to be from OpenShift Origin distribution. `kubectl` v1.7.1 from k8s binary release updates `deployment/hello-node` to new image. I think, it could be just issue with `kubectl` v1.5.2-47-g43a9be4 (see https://github.com/openshift/origin/blob/v1.5.1/Godeps/Godeps.json#L5433-L5487).

I think, issue could be closed since it's not reproducible on v1.7.1
